### PR TITLE
[Feature] - API Middleware

### PR DESCRIPTION
**Pull request by rkapka**
Part of #7510, this PR is a feature branch containing all work required to make the API middleware work as expected.

### CODE REVIEW

**Review by nisdas — APPROVED** [review]
test review

File: beacon-chain/rpc/apimiddleware/endpoint_factory.go
Comment (rkapka) [inline beacon-chain/rpc/apimiddleware/endpoint_factory.go:53]: `Path` is set at the bottom, no need to have it here. Applies to all cases.

File: shared/httputils/endpoint_test.go
Comment (rkapka) [inline shared/httputils/endpoint_test.go:76]: This file should be reverted.

**Review by rauljordan — DISMISSED** [review]
Reviewed once more, giving this a green light. Need one more review before merge

**Review by rauljordan — DISMISSED** [review]

File: shared/bytesutil/bytes_test.go
Comment (prestonvanloon) [inline shared/bytesutil/bytes_test.go:390]: Why remove this test, is this now a valid hex string?

File: beacon-chain/rpc/beaconv1/blocks.go
Comment (prestonvanloon) [inline beacon-chain/rpc/beaconv1/blocks.go:354]: Why is this a pointer receiver? This looks like a conversion or migration method. Is this the most appropriate place to put this method?

File: shared/grpcutils/grpcutils.go
Comment (prestonvanloon) [inline shared/grpcutils/grpcutils.go:87]: Maybe this would be more aptly named as `SetCustomErrorHeader`. Append makes me think I can call it many times, however each time will actually overwrite the previous value. (I think?)

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Partial review. This is a really large PR. Please consider filing smaller, well scoped pull requests to incrementally review/merge these changes.

File: beacon-chain/rpc/nodev1/node.go
Comment (prestonvanloon) [inline beacon-chain/rpc/nodev1/node.go:360]: Please use errors.Wrap

File: shared/gateway/api_middleware_processing.go
Comment (prestonvanloon) [inline shared/gateway/api_middleware_processing.go:26]: Please use errors.Wrap

File: shared/gateway/api_middleware_processing.go
Comment (prestonvanloon) [inline shared/gateway/api_middleware_processing.go:40]: Please use errors.Wrap

File: shared/gateway/api_middleware_processing.go
Comment (prestonvanloon) [inline shared/gateway/api_middleware_processing.go:51]: Please use errors.Wrap

File: shared/gateway/api_middleware_processing.go
Comment (prestonvanloon) [inline shared/gateway/api_middleware_processing.go:158]: Please use errors.Wrap

File: shared/gateway/api_middleware_processing.go
Comment (prestonvanloon) [inline shared/gateway/api_middleware_processing.go:171]: Please use errors.Wrap

File: shared/gateway/api_middleware_processing.go
Comment (prestonvanloon) [inline shared/gateway/api_middleware_processing.go:181]: Please use errors.Wrap

File: shared/gateway/api_middleware_processing.go
Comment (prestonvanloon) [inline shared/gateway/api_middleware_processing.go:208]: Please use errors.Wrap

File: shared/gateway/api_middleware_processing.go
Comment (prestonvanloon) [inline shared/gateway/api_middleware_processing.go:240]: Please use errors.Wrap

File: shared/gateway/api_middleware_processing.go
Comment (prestonvanloon) [inline shared/gateway/api_middleware_processing.go:266]: Please use errors.Wrap

File: beacon-chain/rpc/apimiddleware/custom_handlers.go
Comment (prestonvanloon) [inline beacon-chain/rpc/apimiddleware/custom_handlers.go:26]: Is it spec to call the file name like this? It might be useful to include the slot number in the filename, perhaps.

File: beacon-chain/rpc/apimiddleware/custom_handlers.go
Comment (prestonvanloon) [inline beacon-chain/rpc/apimiddleware/custom_handlers.go:95]: Consider naming to reflect a boolean return type. I.e. `isSSZRequested`.

Also, I think it's the "go way" to always capitalize acronyms. i.e. SSZ, URL, FTP, etc. Consider renaming all `Ssz` to `SSZ`, within reason

File: beacon-chain/rpc/apimiddleware/endpoint_factory.go
Comment (prestonvanloon) [inline beacon-chain/rpc/apimiddleware/endpoint_factory.go:171]: Curious, why does this Endpoint have a path parameter but many other routes do not?

File: beacon-chain/rpc/beaconv1/validator.go
Comment (prestonvanloon) [inline beacon-chain/rpc/beaconv1/validator.go:30]: errors.Wrapf

File: cmd/beacon-chain/flags/base.go
Comment (prestonvanloon) [inline cmd/beacon-chain/flags/base.go:76]: Do you want to add more information to this usage? "API middleware" is kind of vague. Maybe mention this is the endpoint that serves JSON REST calls in accordance with eth2.0-apis.

File: beacon-chain/rpc/beaconv1/blocks.go
Comment (rkapka) [reply beacon-chain/rpc/beaconv1/blocks.go:354]: yeah, it should be in `migration`

File: beacon-chain/rpc/apimiddleware/custom_handlers.go
Comment (terencechain) [inline beacon-chain/rpc/apimiddleware/custom_handlers.go:32]:
```suggestion
func handleGetBeaconBlockSsz(m *gateway.ApiProxyMiddleware, endpoint gateway.Endpoint, writer http.ResponseWriter, request *http.Request) (handled bool) {
```

Future proof when add get shard block later this year

File: beacon-chain/rpc/beaconv1/blocks.go
Comment (terencechain) [inline beacon-chain/rpc/beaconv1/blocks.go:170]: This feels more `Internal` than `InvalidArgument` to me

File: beacon-chain/rpc/beaconv1/pool.go
Comment (terencechain) [inline beacon-chain/rpc/beaconv1/pool.go:215]: I dont think `InvalidArgument` is appropriate here and for the one below

File: beacon-chain/state/stateV0/getters_validator.go
Comment (terencechain) [inline beacon-chain/state/stateV0/getters_validator.go:22]: Not mentioned in the description. Ideally these changes should be its own PR

File: beacon-chain/rpc/apimiddleware/custom_handlers.go
Comment (rkapka) [reply beacon-chain/rpc/apimiddleware/custom_handlers.go:26]: There is no name enforced by the spec. Although I agree that `beacon_state.ssz` is very generic, this function does not have access to any information that could help here. If it's OK, I would like to leave it as it is.

File: beacon-chain/rpc/apimiddleware/endpoint_factory.go
Comment (rkapka) [reply beacon-chain/rpc/apimiddleware/endpoint_factory.go:171]: this is a mistake

File: beacon-chain/rpc/beaconv1/blocks.go
Comment (rkapka) [reply beacon-chain/rpc/beaconv1/blocks.go:170]: Not sure about this one and the other one. `InvalidArgument` will get converted to HTTP status code `400` and https://ethereum.github.io/eth2.0-APIs/#/Beacon/publishBlock says that we should return `400` when
> The SignedBeaconBlock object is invalid

I guess the question is are these errors equivalent to submitting an invalid block?

File: beacon-chain/state/stateV0/getters_validator.go
Comment (rkapka) [reply beacon-chain/state/stateV0/getters_validator.go:22]: I had to create a custom error to be able to discern between different error cases. I agree this should come in a separate PR.

File: beacon-chain/rpc/beaconv1/blocks.go
Comment (terencechain) [reply beacon-chain/rpc/beaconv1/blocks.go:170]: I could see them being both. The error could happen due to 1
1.) Bad input block
2.) Bad HTR implementation 

Although you did use `Internal` for `MarshalSSZ` and they are similar. Perhaps, we can settle into one

Screen Shot 2021-06-14 at 9 07 34 AM

Comment (terencechain) [reply beacon-chain/rpc/beaconv1/blocks.go:170]: Nvm. That block is from the internal DB. That makes sense to me

File: beacon-chain/state/stateV0/getters_validator.go
Comment (rkapka) [reply beacon-chain/state/stateV0/getters_validator.go]: Renamed to `ValidatorIndexOutOfRangeError` to avoid merge conflicts with #9035

File: beacon-chain/rpc/testutil/mock_state_fetcher.go
Comment (prestonvanloon) [inline beacon-chain/rpc/testutil/mock_state_fetcher.go:20]:
```suggestion
// StateRoot --
```

Not that it matters, but yeah... :grin:

File: shared/grpcutils/grpcutils.go
Comment (rkapka) [reply shared/grpcutils/grpcutils.go:87]: No, it will not overwrite the previous value.

> SetHeader sets the header metadata. When called multiple times, all the provided metadata will be merged.

File: beacon-chain/rpc/beaconv1/validator.go
Comment (rkapka) [reply beacon-chain/rpc/beaconv1/validator.go:81]: `||` is correct. See comment:
> Exit early if no matching validators we found **OR** we don't want to further filter validators by status

File: beacon-chain/rpc/testutil/mock_state_fetcher.go
Comment (rkapka) [reply beacon-chain/rpc/testutil/mock_state_fetcher.go:20]: :relieved:

File: beacon-chain/rpc/beaconv1/validator.go
Comment (prestonvanloon) [reply beacon-chain/rpc/beaconv1/validator.go:80]: OK i see.

Can you move the `len(req.Status) == 0 { return nothing }` to the top of the method? There is no need to even query the state if we aren't going to do anything with it.

**Review by prestonvanloon — COMMENTED** [review]
I think this is the last feedback. I've reviewed everything as best as I can. Thanks for resolving the concerns!

**Review by prestonvanloon — APPROVED** [review]
Disregard my last comment. I get it now. Thanks!!

**Pull request merged**
